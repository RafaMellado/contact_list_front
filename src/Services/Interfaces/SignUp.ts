export interface SignUpRequestBody {
  username: string;
  email: string;
  password: string;
  password_confirmation: string;
}

export interface SignUpResponse {
  username: string;
  email: string;
}
