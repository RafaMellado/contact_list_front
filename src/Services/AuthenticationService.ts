import Cookies from "universal-cookie";
import { LoginRequestBody, LoginResponse } from "./Interfaces/Login";
import LoginRepository from "./Repository/LoginRepository";

class AuthenticationService {
  async login(data: LoginRequestBody) {
    const response = await LoginRepository.login(data);

    if (response.token) {
      this.setAuthCookie(response);

      window.location.href = "/home";
    }
  }

  setAuthCookie(data: LoginResponse) {
    const cookies = new Cookies();

    cookies.set(
      "user",
      { token: data.token, username: data.username },
      { expires: new Date(data.exp) }
    );
  }
}

export default new AuthenticationService();
