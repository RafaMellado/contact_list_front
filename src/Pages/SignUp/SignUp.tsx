import { FormEvent, useRef, useState } from "react";
import { Button, Card, Form } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router-dom";
import { ErrorBox } from "../../Components/ErrorBox";
import AuthenticationService from "../../Services/AuthenticationService";
import { SignUpError } from "../../Services/Interfaces/SignUp";

export function SignUp() {
  const { t } = useTranslation();
  const navigate = useNavigate();
  const [errors, setErrors] = useState<SignUpError>();

  const TRANSLATIONS = "signUp";

  const username = useRef<HTMLInputElement>(null);
  const email = useRef<HTMLInputElement>(null);
  const password = useRef<HTMLInputElement>(null);
  const passwordConfirmation = useRef<HTMLInputElement>(null);

  const backToLogin = () => {
    navigate("/");
  };

  const submit = async (event: FormEvent<HTMLFormElement>) => {
    event?.preventDefault();

    try {
      await AuthenticationService.signUp({
        username: String(username?.current?.value),
        email: String(email?.current?.value),
        password: String(password?.current?.value),
        password_confirmation: String(passwordConfirmation?.current?.value),
      });

      backToLogin();
    } catch (errors) {
      setErrors(errors as SignUpError);
    }
  };

  return (
    <div className="d-flex justify-content-center">
      <Card>
        <Card.Body>
          <Form onSubmit={submit}>
            <Form.Group className="mb-3">
              <Form.Label>{t<string>(`${TRANSLATIONS}.username`)}</Form.Label>
              <Form.Control
                as="input"
                type="text"
                data-testid="sign-up-username"
                placeholder={t<string>(`${TRANSLATIONS}.username`)}
                minLength={6}
                maxLength={20}
                required
                ref={username}
              />
            </Form.Group>

            <Form.Group className="mb-3">
              <Form.Label>{t<string>(`${TRANSLATIONS}.email`)}</Form.Label>
              <Form.Control
                as="input"
                type="email"
                data-testid="sign-up-email"
                placeholder={t<string>(`${TRANSLATIONS}.email`)}
                required
                ref={email}
              />

              {errors && <ErrorBox errors={errors.errors} field="email" />}
            </Form.Group>

            <Form.Group className="mb-3">
              <Form.Label>{t<string>(`${TRANSLATIONS}.password`)}</Form.Label>
              <Form.Control
                as="input"
                type="password"
                data-testid="sign-up-password"
                placeholder={t<string>(`${TRANSLATIONS}.password`)}
                minLength={6}
                required
                ref={password}
              />
            </Form.Group>

            <Form.Group className="mb-3">
              <Form.Label>
                {t<string>(`${TRANSLATIONS}.passwordConfirmation`)}
              </Form.Label>
              <Form.Control
                as="input"
                type="password"
                data-testid="sign-up-password-confirmation"
                placeholder={t<string>(`${TRANSLATIONS}.passwordConfirmation`)}
                required
                ref={passwordConfirmation}
              />

              {errors && (
                <ErrorBox
                  errors={errors.errors}
                  field="password_confirmation"
                />
              )}
            </Form.Group>

            <div className="d-flex justify-content-between">
              <Button
                className="w-100"
                variant="secondary"
                onClick={backToLogin}
              >
                {t<string>(`${TRANSLATIONS}.back`)}
              </Button>
              <Button
                data-testid="sign-up-submit"
                className="ms-2 w-100"
                variant="primary"
                type="submit"
              >
                {t<string>(`${TRANSLATIONS}.submit`)}
              </Button>
            </div>
          </Form>
        </Card.Body>
      </Card>
    </div>
  );
}
