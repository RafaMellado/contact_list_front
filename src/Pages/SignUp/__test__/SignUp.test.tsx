import { SignUp } from "../SignUp";
import { act, fireEvent, render } from "@testing-library/react";
import { Router } from "react-router-dom";
import { useState as useStateMock } from "react";
import AuthenticationService from "../../../Services/AuthenticationService";
import { createMemoryHistory } from "history";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock("../../../Services/AuthenticationService", () => ({
  signUp: jest.fn(),
}));

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

const setState = jest.fn();

describe(`<${SignUp.name} />`, () => {
  const history = createMemoryHistory();

  const factoryComponent = () => {
    return render(
      <Router location="/" navigator={history}>
        <SignUp />
      </Router>
    );
  };

  beforeEach(() => {
    (useStateMock as jest.Mock).mockImplementation(() => [null, setState]);
  });

  test("renders correctly the component", () => {
    const component = factoryComponent();

    act(() => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  test("fill correctly and call signUp", () => {
    const component = factoryComponent();

    act(() => {
      fireEvent.input(component.getByTestId("sign-up-username"), {
        target: {
          value: "abcdef",
        },
      });

      fireEvent.input(component.getByTestId("sign-up-email"), {
        target: {
          value: "test@email.com",
        },
      });

      fireEvent.input(component.getByTestId("sign-up-password"), {
        target: {
          value: "abcdef",
        },
      });

      fireEvent.input(component.getByTestId("sign-up-password-confirmation"), {
        target: {
          value: "abcdef",
        },
      });

      fireEvent.click(component.getByTestId("sign-up-submit"));

      expect(AuthenticationService.signUp).toHaveBeenCalled();
    });
  });
});
