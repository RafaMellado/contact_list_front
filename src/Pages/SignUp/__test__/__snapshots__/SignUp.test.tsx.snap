// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center"
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              signUp.username
            </label>
            <input
              class="form-control"
              data-testid="sign-up-username"
              maxlength="20"
              minlength="6"
              placeholder="signUp.username"
              required=""
              type="text"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              signUp.email
            </label>
            <input
              class="form-control"
              data-testid="sign-up-email"
              placeholder="signUp.email"
              required=""
              type="email"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              signUp.password
            </label>
            <input
              class="form-control"
              data-testid="sign-up-password"
              minlength="6"
              placeholder="signUp.password"
              required=""
              type="password"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              signUp.passwordConfirmation
            </label>
            <input
              class="form-control"
              data-testid="sign-up-password-confirmation"
              placeholder="signUp.passwordConfirmation"
              required=""
              type="password"
            />
          </div>
          <div
            class="d-flex justify-content-between"
          >
            <button
              class="w-100 btn btn-secondary"
              type="button"
            >
              signUp.back
            </button>
            <button
              class="ms-2 w-100 btn btn-primary"
              data-testid="sign-up-submit"
              type="submit"
            >
              signUp.submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
